### PR TITLE
Boot: Restore device's last location if opened at /

### PR DESCRIPTION
**Pull request by mcsf**
See #11755

# Gist

- If the initial path is `/`, look for the last known path and start from there. The path is intentionally not persisted server-side, as this is a per-device behavior. See Desktop app for a precedent.
- Done as a lightning experiment at ∆'s meetup; let's try to merge it quickly.
- Enabled in `development` and `wpcalypso`.

remember-location

# Open questions

- [x] ~`route-intercept` is not the greatest name~ `restore-last-location` adopted
- [ ] specific objections to use of `store`?

### CODE REVIEW

**Comment by bisko** [comment]
The PR does exactly what it needs to and is working awesomely.

There's a weird button state change while loading:

buttons-change

The `Reader` button stays highlighted until everything loads fully. Then it switches to `My Sites`.

This doesn't seem like a blocker for merging the current PR.

**Comment by jblz** [comment]
Code looks good -- I like how self contained it is since you moved it to the module & the feature flag will help us see how this feels in a more realistic context.

Can you please stub out a `client/state/routing/README.md` so we document in plain language the evolution of our thinking here?

For others reviewing this from a usability standpoint, it's worth mentioning that the Desktop app [behaves quite similarly](https://github.com/Automattic/wp-desktop/blob/329205c38befd21695c73a57c10dd49d5359d8f8/desktop/server/index.js#L29-L72) to this proposed change.

**Comment by mcsf** [comment]
Final thumb? :kiss:

File: client/state/routing/middleware.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Internal dependencies
+ */
+import { ROUTE_SET } from 'state/action-types';
+
+const debug = debugFactory( 'calypso:restore-last-location' );
+
+let hasInitialized = false;
+
+export const restoreLastLocation = () => ( next ) => ( action ) => {
```
Comment (gwwar) [inline client/state/routing/middleware.js]: @mcsf Isn't this signature normally `export const restoreLastLocation = store => next => action => {`

Comment (gwwar) [inline client/state/routing/middleware.js]: Ah I see, any reason why we're not using `localforage` directly then?

Comment (mcsf) [inline client/state/routing/middleware.js]: Only my ignorance. :) I looked at prior art and found `store`, but I can switch to LF.

> Isn't this signature normally

Yes, but `store` was never used.

Comment (gwwar) [inline client/state/routing/middleware.js]: Right 😄  I was just confused earlier because of `store`name matching the redux `store`

**Pull request merged**